aesd-char-driver/main.c

```diff
@@ -181,12 +182,112 @@ ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
 
     return retval;
 }
+
+loff_t aesd_llseek(struct file *filp,loff_t offset,int whence)
+{
+   int len=0,i;
+   loff_t retval;
+   struct aesd_dev *dev = filp->private_data;
+   
+   switch(whence)
+   {
+   
+   	case SEEK_CUR:
+   		retval=filp->f_pos+offset;
+   		break;
+   	case SEEK_SET:
+   		retval=filp->f_pos;
+   		break;
+   	case SEEK_END:
+   		if(mutex_lock_interruptible(&dev->lock)!=0)
+   		{
+   			retval=-ERESTART;
+   			goto out;
+   		}
+   		for(i=dev->buffer.out_offs;i!=dev->buffer.in_offs;i = (i + 1) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED)
+   		{
+   		 len+=dev->buffer.entry[i].size;
+   		}
+   		mutex_unlock(&dev->lock);
+   		retval=len+offset;
+   		break;
+   	default:
+   		retval=-EINVAL;
+   		goto out;
+   	
+   		
+   }
+   if(retval<0)
+   {
+   	return -EINVAL;
+   }
+   filp->f_pos=retval;
+   out:
+   	return retval;
+}
+
+long aesd_unlocked_ioctl(struct file *filp,unsigned int cmd, unsigned long arg)
+{
+	struct aesd_dev *device = filp->private_data;
+	struct aesd_seekto seek_params;
+    	int accumulated_length = 0;
+    	long result = 0;
+    	// Validate command type
+    	if (cmd != AESDCHAR_IOCSEEKTO) 
+    	{
+        	printk(KERN_ERR "Invalid IOCTL command received\n");
+        	return -ENOTTY;
+    	}
+    	// Copy data from user space
+    	if (copy_from_user(&seek_params, (struct aesd_seekto __user *)arg, sizeof(seek_params)) != 0) 
+    	{
+        	printk(KERN_ERR "Failed to copy data from user\n");
+        	return -EFAULT;
+    	}
+    	 // Check if command index is within bounds
+    	if (seek_params.write_cmd >= AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED)
+    	{
+        	printk(KERN_ERR "Command index out of range\n");
+        	return -EINVAL;
+    	}
+    	// Calculate the actual position in the circular buffer
+    	seek_params.write_cmd = (seek_params.write_cmd + device->buffer.out_offs) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
```

Comment:
@shruthi-thallapally what does it mean to access the device buffer here and in line 257 without holding the mutex?

Reply:
@dwalkes I missed that part. I'll be more careful next time.